### PR TITLE
Enable saving and loading stateful DataLoaders in Trainer

### LINKED ISSUE #17105: Support manual dataloader fault-tolerance
### Description & Motivation

The PyTorch Lightning 2.0 release removed support for the experimental fault tolerance feature because its implementation was too complicated and added complexity across the codebase.

This is because its scope was too broad, trying to automatically add fault-tolerance to the regular `DataLoader`, with abstractions leaking across the Trainer and loops.

However, this is still a valuable feature that we want to enable. We need to find the right way to do it.

### Pitch

Since the `CombinedLoader` is always used after 2.0, we can have it implement the `Stateful` interface https://github.com/Lightning-AI/lightning/blob/e9937476a2c54f764e786a97f00f0f4baff74919/src/lightning/fabric/utilities/types.py#L43-L51

So that it serializes/deserializes its state, the state of its iterator mode, and the state of any iterables that also define the stateful interface.

And include this state in the checkpoint.

Additionally, If we wanted to support fault-tolerance for the `torch.utils.data.DataLoader`, we could write our wrapper that implements it and ask the user to use it manually. This should encapsulate the logic well.
We could also not do this part and have people rely on stateful dataloader implementations such as [`DataLoader2`](https://pytorch.org/data/beta/dataloader2.html#torchdata.dataloader2.DataLoader2.state_dict)

If done correctly, this feature could be used with `Fabric` too, as the `CombinedLoader` is external to the Trainer.

### Alternatives

Don't do this.

### Additional context

Support for this feature was removed in #16533

Integrations like `StreamingDataset` expose this: https://github.com/mosaicml/streaming/blob/da53bc5911d012ce37fdf6eb41165d6eae07bc92/streaming/base/dataset.py#L600-L648

cc @borda @justusschock @awaelchli @carmocca

### LINKED ISSUE #17543: Resuming training gives different model result / weights
### Expected behavior

Saving a checkpoint and resuming from that checkpoint in Pytorch Lightning (2.0.2) should give the same model in the end (weights, states, loss).

### Actual behavior

Saving a checkpoint and resuming from that checkpoint in Pytorch Lightning (2.0.2) gives different training result in the end (checked via MD5, loss etc). This brakes continual learning.

### What version are you seeing the problem on?

v2_0

### How to reproduce the bug

```python
import os
import warnings

warnings.filterwarnings("ignore")  # avoid printing out absolute paths

import copy
from pathlib import Path
import warnings
import tempfile
import json, hashlib

import numpy as np
import pandas as pd
import torch
import random
import lightning.pytorch  as pl

from pytorch_forecasting import TimeSeriesDataSet, NHiTS

import logging
logging.getLogger("lightning.pytorch.accelerators.cuda").setLevel(logging.WARNING)

##################### Eval

class NHitsRunner():

	def __init__(self):

		self.batch_size = 256 
		self.reset_seed()
		self.create_datasets()
		self.create_model()
		self.path = None

	def create_model(self):

		# Dataset

		self.nhits_model = NHiTS.from_dataset(
			self.test_datasets[0],
			learning_rate=0.1,
			hidden_size=16,
			optimizer='adam'
		)

		self.trainer = pl.Trainer(
			devices=1, accelerator="gpu",
			max_epochs=1,
			gradient_clip_val=0.1,
			enable_model_summary=False,
			enable_progress_bar=True,
			enable_checkpointing=False,
			logger=False,
		)

	def create_dataset(self, frame):

		self.max_prediction_length = 10
		self.max_encoder_length = 24
		training_cutoff = frame["time_idx"].max() - self.max_prediction_length

		return TimeSeriesDataSet(
			frame[lambda x: x.time_idx <= training_cutoff],
			time_idx="time_idx",
			target="price",
			group_ids=["group_id"],
			max_encoder_length=self.max_encoder_length,
			max_prediction_length=self.max_prediction_length,
			static_categoricals=[],
			time_varying_known_reals=[],
			time_varying_unknown_categoricals=[],
			time_varying_unknown_reals=[
				"price",
				"volume",
			],
			target_normalizer=None,
		)

	def train_epoch(self, set_id):

		train_dataloader = self.test_datasets[set_id].to_dataloader(train=True, batch_size=self.batch_size, num_workers=0)

		if self.path != None:

			print('Consumed trainer checkpoint')

			self.trainer.fit(
				self.nhits_model,
				ckpt_path=self.path,
				train_dataloaders=train_dataloader,
			)

			os.remove(self.path)

			self.path = None

		else:
			
			self.trainer.fit(
				self.nhits_model,
				train_dataloaders=train_dataloader,
			)

		self.trainer.fit_loop.max_epochs += int(1)

	def reset_seed(self):

		pl.seed_everything(0)
		torch.set_float32_matmul_precision('medium')

	def create_frame(self):

		# Create sine wave
		dataframe_size = 128 * 64
		test_dataframe = pd.DataFrame({'time_idx': np.arange(0, dataframe_size, 1, dtype=int)})
		test_dataframe = test_dataframe.assign(price=lambda x: np.sin(x.time_idx/3)*100)
		test_dataframe = test_dataframe.assign(volume=lambda x: np.sin(x.time_idx/2)*200)
		test_dataframe['group_id'] = 'group_a'
		return test_dataframe

	def create_datasets(self):

		self.test_frames = []
		self.test_datasets = []

		for i in range(3):

			new_frame = self.create_frame()

			self.test_frames.append(new_frame)

			self.test_datasets.append(self.create_dataset(new_frame))

	def convert_dict(self, d, flat_dict):
		for k, v in d.items():
			if isinstance(v, dict):
				self.convert_dict(v, flat_dict)
			else:
				if isinstance(v, torch.Tensor):
					v = v.cpu().detach().numpy().tolist()
				flat_dict[k] = v

	def dict_md5(self):

		_dict = self.nhits_model.state_dict()
		
		flat_dict = {}
		self.convert_dict(_dict, flat_dict)

		data_md5 = hashlib.md5(json.dumps(flat_dict, sort_keys=True).encode('utf-8')).hexdigest()

		return data_md5

	def test_save_load_trainer_checkpoint(self):

		print("Saving trainer checkpoint")

		# Save trainer file

		model_filename = "test_checkpoint_forecast.ckpt"
		tempdir = tempfile.gettempdir()
		file_name = tempdir+"/"+model_filename
		self.trainer.save_checkpoint(file_name)

		self.path = file_name

#### Test running WITHOUT saving and loading 

runner_1 = NHitsRunner()

for i in range(3):

	runner_1.train_epoch(i)

run1_md5 = runner_1.dict_md5()

#### Test running WITHOUt saving and loading 

runner_2 = NHitsRunner()

for i in range(3):

	if i == 2:
		runner_2.test_save_load_trainer_checkpoint()

	runner_2.train_epoch(i)

run2_md5 = runner_2.dict_md5()

#### Compare

if run1_md5 != run2_md5:

	print('Models md5 differs {} != {}'.format(run1_md5, run2_md5))

else:

	print('Models md5 equals {} == {}'.format(run1_md5, run2_md5))
```

### Error messages and logs

```
Global seed set to 0
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.52it/s, train_loss_step=0.737, train_loss_epoch=1.700]`Trainer.fit` stopped: `max_epochs=1` reached.
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.51it/s, train_loss_step=0.737, train_loss_epoch=1.700]
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.75it/s, train_loss_step=0.648, train_loss_epoch=1.150]`Trainer.fit` stopped: `max_epochs=2` reached.
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.74it/s, train_loss_step=0.648, train_loss_epoch=1.150]
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.43it/s, train_loss_step=0.598, train_loss_epoch=1.440]`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.41it/s, train_loss_step=0.598, train_loss_epoch=1.440]
Global seed set to 0
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.57it/s, train_loss_step=0.737, train_loss_epoch=1.700]`Trainer.fit` stopped: `max_epochs=1` reached.
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.55it/s, train_loss_step=0.737, train_loss_epoch=1.700]
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.70it/s, train_loss_step=0.648, train_loss_epoch=1.150]`Trainer.fit` stopped: `max_epochs=2` reached.
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.69it/s, train_loss_step=0.648, train_loss_epoch=1.150]
Saving trainer checkpoint
Consumed trainer checkpoint
Restoring states from the checkpoint path at /tmp/test_checkpoint_forecast.ckpt
Restored all states from the checkpoint at /tmp/test_checkpoint_forecast.ckpt
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.91it/s, train_loss_step=0.342, train_loss_epoch=0.470]`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.90it/s, train_loss_step=0.342, train_loss_epoch=0.470]
Models md5 differs 0d6210c3c296c7acfccac4a4b27e1a16 != 25eee25731cd9d21e9991afa47a90b82

```

### Environment

<details>
  <summary>Current environment</summary>

```
- PyTorch-Forecasting version: 1.0.0 (based on LightningModule)
- PyTorch version: 2.0.0
- PyTorch lightning: 2.0.2 
- Python version: 3.8.10
- Operating System: Ubuntu 20.04
- CUDA 12.1
- Running environment: Local
```

</details>

### More info

In the example code I create 3 data frames and train on each, one epoch at the time. This will render a md5 sum. Rerunning the code will render the same md5 sum.

The second time I run the exact same code, before the last iteration, I save a checkpoint to disk and then resume training from that checkpoint. This will render a different md5 sum. This should not be the case.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Saves the `state_dict` of training dataloaders into the checkpoint, and enables that state to be loaded when resuming from a checkpoint. An example of a stateful dataloader is `lightning.data.StreamingDataLoader`. 

The implementation collects the state of all iterables under the CombinedLoader that follow the stateful interface. The states are collected over the flattened view (CombinedLoader.flattened) and stored in a last. They get restored in the same manner via the flattened view. This means the number of iterables and the order in which they are given must be exactly the same as when the checkpoint was saved, otherwise the loading will fail. 

Fixes #17105
Closes #17543

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Nice ;)

File: src/lightning/pytorch/loops/fit_loop.py [hunk 1]
```diff
@@ -409,9 +414,34 @@ def teardown(self) -> None:
...
     def _iteration_based_training(self) -> bool:
         return self.trainer.max_steps != -1
+
+    def _restore_combined_loader_state(self) -> None:
+        if not self._combined_loader_states:
+            return
+
+        loaders = self._combined_loader.flattened if self._combined_loader is not None else []
+        stateful_loaders = [loader for loader in loaders if isinstance(loader, _Stateful)]
+        for loader, state_dict in zip(stateful_loaders, self._combined_loader_states):
```
Comment (carmocca) [inline src/lightning/pytorch/loops/fit_loop.py:444]: zip will allow that stateful_loaders and combined_loader_states have different lengths. It's probably safer to raise an error if they have different lengths

File: src/lightning/pytorch/loops/fit_loop.py [hunk 2]
```diff
@@ -94,6 +95,7 @@ def __init__(
 
         self._data_source = _DataLoaderSource(None, "train_dataloader")
         self._combined_loader: Optional[CombinedLoader] = None
+        self._combined_loader_states: Optional[List[Dict[str, Any]]] = None
```
Comment (carmocca) [inline src/lightning/pytorch/loops/fit_loop.py:98]: Having this default to an empty list should be equivalent and avoid the typing issues around None checks

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/loops/fit_loop.py:98]: We need to delay loading the states until the combined loader gets initialized in the loop's setup_data. We need to store it somewhere until then. I thought the loop here would be the best place. I don't think there is a way to store it inside the loader. However, I could add methods to the CombinedLoader to save and restore a list of states.

I could rename the variable to something more obvious, like `_combined_loader_states_to_load`.

File: src/lightning/pytorch/loops/fit_loop.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/pytorch/loops/fit_loop.py]: Good idea, I added the error checking

File: src/lightning/pytorch/loops/fit_loop.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/loops/fit_loop.py]: Based on your comments, I did a refactor adding `CombinedLoader._state_dicts()` and `._load_state_dicts()`, moving logic for gathering the dicts into the combined loader. I like this better.

File: src/lightning/pytorch/utilities/combined_loader.py [hunk 3]
```diff
@@ -374,6 +375,24 @@ def _dataset_length(self) -> int:
         fn = _SUPPORTED_MODES[self._mode]["fn"]
         return fn(lengths)
 
+    def _state_dicts(self) -> List[Dict[str, Any]]:
+        """Returns the list of state dicts for iterables in `self.flattened` that are stateful."""
+        return [loader.state_dict() for loader in self.flattened if isinstance(loader, _Stateful)]
+
+    def _load_state_dicts(self, states: List[Dict[str, Any]]) -> None:
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/utilities/combined_loader.py:382]: Mainly because these are lists, and so the types don't match the `_Stateful` protocol. It seemed the right thing to do given that we use the `.flattened` representation (I opted for simplicity here). Would a different format make more sense?

[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/utilities/combined_loader.py:382]: Good point. We could use the pytree spec to reconstruct the state dicts in the structure of the iterables, meaning returning a non-flattened view of the state dicts.

But we don't have a use for that, and nobody has requested this, so happy to keep it flattened and protected for now
